File: haskell/glue/src/Glue/Lib/List/Partition.hs
module Glue.Lib.List.Partition where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception
import Glue.IR (IR (..))

partition :: IR Eval
partition = NativeFunc partitionImpl

partitionImpl :: IR Eval -> Eval (IR Eval)
partitionImpl predicateIR = pure $ NativeFunc (partitionList predicateIR)

partitionList :: IR Eval -> IR Eval -> Eval (IR Eval)
partitionList predicateIR listIR = do
    predicate <- eval predicateIR
    list <- eval listIR
    case list of
        List xs -> do
            -- Partition list into two lists based on predicate
            (matching, nonMatching) <- partitionElements predicate xs
            pure $ List [List matching, List nonMatching]
        _ -> throwError $ wrongArgumentType ["function", "list"]

-- Helper function to partition list based on predicate
partitionElements :: IR Eval -> [IR Eval] -> Eval ([IR Eval], [IR Eval])
partitionElements _ [] = pure ([], [])
partitionElements predicate (x : xs) = do
    -- Evaluate (predicate x) and check if it returns true
    result <- eval (List [predicate, x])
    (matching, nonMatching) <- partitionElements predicate xs
    case result of
        Bool True -> pure (x : matching, nonMatching)
        Bool False -> pure (matching, x : nonMatching)
        _ -> throwError $ wrongArgumentType ["boolean result from predicate"]

File: dart/glue/lib/src/lib/list/partition.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Partition function - splits list into two lists based on predicate
/// Mirrors Haskell Glue.Lib.List.Partition.partition exactly
Ir partition = IrNativeFunc(partitionImpl);

/// Partition function implementation
/// Mirrors Haskell Glue.Lib.List.Partition.partitionImpl exactly
Eval<Ir> partitionImpl(Ir predicateIr) {
  return Eval.pure(IrNativeFunc(partitionList(predicateIr)));
}

/// Helper function for list argument
/// Mirrors Haskell Glue.Lib.List.Partition.partitionList exactly
Eval<Ir> Function(Ir) partitionList(Ir predicate) {
  return (Ir list) {
    return switch (list) {
      IrList(elements: final elements) =>
        partitionElements(predicate, elements.toList()).map((partitioned) {
          final (matching, nonMatching) = partitioned;
          return IrList([IrList(matching), IrList(nonMatching)]);
        }),
      _ => throwError(wrongArgumentType(['function', 'list'])),
    };
  };
}

/// Helper function to partition list based on predicate
Eval<(List<Ir>, List<Ir>)> partitionElements(Ir predicate, List<Ir> elements) {
  if (elements.isEmpty) {
    return Eval.pure(([], []));
  }

  return applyPredicate(predicate, elements[0]).flatMap((satisfies) {
    return partitionElements(predicate, elements.sublist(1)).map((partitioned) {
      final (matching, nonMatching) = partitioned;
      if (satisfies) {
        return ([elements[0], ...matching], nonMatching);
      } else {
        return (matching, [elements[0], ...nonMatching]);
      }
    });
  });
}

/// Helper function to apply predicate to an element
Eval<bool> applyPredicate(Ir predicate, Ir element) {
  return eval(IrList([predicate, element])).flatMap((result) {
    if (result is IrBool) {
      return Eval.pure(result.value);
    } else {
      return throwError(wrongArgumentType(['boolean result from predicate']));
    }
  });
}
